system/free_tcp_port.go

```diff
@@ -3,16 +3,50 @@ package system
 import (
 	"net"
 	"strconv"
+	"errors"
 )
 
-func FindFreePort() (int, error) {
-	l, _ := net.Listen("tcp", ":0")
-	defer l.Close()
+const (
+	MIN_ACCEPTED_PORT int = 1024
+	MAX_ACCEPTED_PORT int = 65535
+)
 
-	parsedPort, parseErr := strconv.ParseInt(l.Addr().String()[5:], 10, 32)
-	if parseErr != nil {
-		return -1, parseErr
+type FreeTcpPort interface {
+	FindFreePortInRange(minport int, maxport int) (int, error)
+}
+type FreeRangeTcpPort struct {
+	FreeTcpPort
+	IsPortAvailable func(num int) bool
+}
+
+func NewFreeTcpPort() FreeTcpPort {
+	return &FreeRangeTcpPort{IsPortAvailable: isPortAvailable}
+}
+func isPortAvailable(num int) bool {
+	l, err := net.Listen("tcp", ":" + strconv.Itoa(num))
+	if err != nil {
+		return false
 	}
+	l.Close()
+	return true
+}
 
-	return int(parsedPort), nil
+func (f FreeRangeTcpPort) FindFreePortInRange(minport int, maxport int) (int, error) {
+	if minport > maxport {
+		return -1, errors.New("Not valid range port: minimum port is higher than maximum port")
+	}
+	if minport < MIN_ACCEPTED_PORT {
+		return -1, errors.New("Not valid range port: minimum port is lower than " + strconv.Itoa(MIN_ACCEPTED_PORT))
+	}
+	if maxport > MAX_ACCEPTED_PORT {
+		return -1, errors.New("Not valid range port: maximum port is higher than " + strconv.Itoa(MIN_ACCEPTED_PORT))
+	}
+	port := minport
+	for port < maxport {
```

Comment:
Minor case remaining to cover when minport=maxport, it should return a single port if available.